traits/ctraits.c

```diff
@@ -2908,22 +2897,31 @@ static trait_setattr setattr_handlers[] = {
     /*  End of __setstate__ method entries */
     NULL};
 
-static int
-trait_init(trait_object *trait, PyObject *args, PyObject *kwds)
+
+trait_object *
+trait_new (PyTypeObject *trait_type, PyObject *args, PyObject *kw)
 {
-    int kind;
+    int kind, n;
+    trait_object *trait;
 
-    if (!PyArg_ParseTuple(args, "i", &kind)) {
-        return -1;
+    n = PyTuple_GET_SIZE(args);
+
+    if (n == 0){
+        kind = 0;
+    }
+    else if (!PyArg_ParseTuple(args, "i", &kind)) {
+        return NULL;
     }
 
     if ((kind >= 0) && (kind <= 8)) {
+        trait = (trait_object*) PyType_GenericNew(trait_type, args, kw);
         trait->getattr = getattr_handlers[kind];
         trait->setattr = setattr_handlers[kind];
-        return 0;
+        return trait;
     }
 
-    return bad_trait_error();
+    PyErr_SetString(TraitError, "Invalid argument to trait constructor.");
```

Comment:
short exception messages shouldn't usually include a capital letter or period; let's take this opportunity to fix the message
```suggestion
    PyErr_SetString(TraitError, "invalid argument to trait constructor");
```